libraries/AP_Quicktune/AP_Quicktune.h

```diff
@@ -0,0 +1,190 @@
+    AP_Float double_time;
+    AP_Float gain_margin;
+    AP_Float osc_smax;
+    AP_Float yaw_p_max;
+    AP_Float yaw_d_max;
+    AP_Float rp_pi_ratio;
+    AP_Float y_pi_ratio;
+    AP_Int8 auto_filter;
+    AP_Float auto_save;
+    AP_Float max_reduce;
+    AP_Int16 options;
+
+    float get_time() {
+        return AP_HAL::millis() * 0.001;
+    }
+
+    enum class axis_names : uint8_t {
+        RLL = 0,
+        PIT = 1,
+        YAW = 2,
+        DONE = 3,
+        END = 4,
+    };
+
+    enum class param_s : uint8_t {
+        RLL_P,
+        RLL_I,
+        RLL_D,
+        RLL_SMAX,
+        RLL_FLTT,
+        RLL_FLTD,
+        RLL_FLTE,
+        RLL_FF,
+        PIT_P,
+        PIT_I,
+        PIT_D,
+        PIT_SMAX,
+        PIT_FLTT,
+        PIT_FLTD,
+        PIT_FLTE,
+        PIT_FF,
+        YAW_P,
+        YAW_I,
+        YAW_D,
+        YAW_SMAX,
+        YAW_FLTT,
+        YAW_FLTD,
+        YAW_FLTE,
+        YAW_FF,
+        END,
+    };
+
+    //Also the gains
+    enum class stages : uint8_t {
+        D,
+        P,
+        DONE,
+        I,
+        FF,
+        SMAX,
+        FLTT,
+        FLTD,
+        FLTE,
+        END,
+    };
+
+    stages current_stage = stages::D;
+    float last_stage_change = get_time();
+    float last_gain_report = get_time();
+    float last_pilot_input = get_time();
+    float tune_done_time = 0; //nil
+    param_s slew_parm = param_s::END; //nil
+    float slew_target = 0;
+    uint8_t slew_steps = 0;
+    float slew_delta = 0;
+    float last_update = 0;
+
+    uint32_t axes_done = 0;
+    uint32_t filters_done = 0;
+
+    //Bitmask of changed parameters
+    uint32_t param_changed = 0; //{}
+
+    //Saved values of the parameter
+    float param_saved[uint8_t(param_s::END)];
+
+    bool need_restore = false;
+
+    uint32_t last_warning = get_time();
+
+    void reset_axes_done();
+    void setup_filters(axis_names axis);
+    bool have_pilot_input();
+    axis_names get_current_axis();
+    float get_slew_rate(axis_names axis);
+    void advance_stage(axis_names axis);
+    void adjust_gain(param_s param, float value);
+    void adjust_gain_limited(param_s param, float value);
+    float get_gain_mul();
+    void restore_all_params();
+    void save_all_params();
+    bool item_in_bitmask(uint8_t item, uint32_t bitmask);
+    param_s get_pname(axis_names axis, stages stage);
+    AP_Float *get_param_pointer(param_s param);
+    float get_param_value(param_s param);
+    void set_param_value(param_s param, float value);
+    void set_and_save_param_value(param_s param, float value);
+    float gain_limit(param_s param);
+    axis_names get_axis(param_s param);
+    float limit_gain(param_s param, float value);
+    const char* get_param_name(param_s param);
+    void set_bitmask(bool value, uint32_t &bitmask, uint8_t position);
+    stages get_stage(param_s param);
+    const char* get_axis_name(axis_names axis);
+
+    AP_Arming *arming = AP::arming().get_singleton();
```

Comment:
AP_Arming &arming = *AP::arming().get_singleton();
?

Reply:
Changed to inline method, as discussed.